Client/Frontend/Home/TopSitesPanel.swift

```diff
@@ -213,14 +213,19 @@ private class TopSitesCollectionView: UICollectionView {
 
 private class TopSitesLayout: UICollectionViewLayout {
     private var thumbnailRows: Int {
-        return max(2, Int((self.collectionView?.frame.height ?? self.thumbnailHeight) / self.thumbnailHeight))
+        return max(2, Int((self.height ?? self.thumbnailHeight) / self.thumbnailHeight))
     }
 
     private var thumbnailCols = 2
     private var thumbnailCount: Int {
         return thumbnailRows * thumbnailCols
     }
-    private var width: CGFloat { return self.collectionView?.frame.width ?? 0 }
+
+    private lazy var width: CGFloat = { return self.collectionView?.frame.width ?? 0 }()
+
+    private lazy var height: CGFloat? = {
+        return self.collectionView?.frame.height
+        }()
```

Comment:
The problem with that is that it defeats the point of having those values.
We need to be able to explicitly set them otherwise we use the wrong
height/width when we rotate (which was the cause of this bug in the first
place).

I'll have a look to see if I can come up with a solution that solves both
cases.

On 14 August 2015 at 18:42, Stephan Leroux <notifications@github.com> wrote:

> In Client/Frontend/Home/TopSitesPanel.swift
> <https://github.com/mozilla/firefox-ios/pull/932#discussion_r37100102>:
>
> >      }
> >
> >      private var thumbnailCols = 2
> >      private var thumbnailCount: Int {
> >          return thumbnailRows * thumbnailCols
> >      }
> > -    private var width: CGFloat { return self.collectionView?.frame.width ?? 0 }
> > +
> > +    private lazy var width: CGFloat = { return self.collectionView?.frame.width ?? 0 }()
> > +
> > +    private lazy var height: CGFloat? = {
> > +        return self.collectionView?.frame.height
> > +        }()
>
> Ah I think I know the problem with the extra row at the bottom. It looks
> like viewDidLoad calls height/width and at that point the collectionView's
> frame hasn't been updated by autolayout yet so it reports a height of 667
> which will result in 5 rows. Since these are lazy, height/width will only
> evaluate once and cache that result. Instead of being lazy, these should be
> computed properties so we don't cache the height/width values
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mozilla/firefox-ios/pull/932/files#r37100102>.
>